### PR TITLE
Add streaming impl changes + Logical pattern impl + Testcase fixes

### PR DESCRIPTION
**Pull request by mohanvive**
## Purpose

- To fix the compiler and cpu level issues due to the latest language changes.
- Fixes to stream test cases to align with the latest language changes
- Add initial implementation of patterns in streaming for logical patterns

Logical patterns match events that arrive in temporal order and correlate them with logical relationships such as and, or and not. This PR contains the implementation of "or" along with a test case. With this, users can define streaming queries containing logical patterns with relationships made using "or".
Below is a sample query in ballerina.

```
from  every regulatorStateChangeStream where userAction == "on" as e1
        followed by roomKeyStream where e1.roomNo==roomNo && userAction == "removed" as e2
        || regulatorStateChangeStream where e1.roomNo == roomNo && userAction == "off" as e3
        select e1.roomNo as roomNo, e2 == null ? "none" : "stop" as userAction having userAction != "none"  => (RoomKeyAction [] keyAction) {
            regulatorActionStream.publish(keyAction);
        }
```
Here we match a pattern where the regulator is switched on, then check whether the room key was removed or regulator is switched off. in the latter case, where the regulator is switched off as the second event, we don't need to trigger an alert. We only need to trigger an alert if the regulator is first switched on then the room key was removed ( means that the room is closed without switching off the regulator)

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -516,17 +517,16 @@ public void visit(BLangBlockStmt block) {
...
-            ((BLangRecordLiteral) varDefNode.var.expr).name = varDefNode.var.name;
-        }
-
         varDefNode.var = rewrite(varDefNode.var, env);
         BLangVariable varNode = varDefNode.var;
 
         // Generate default init expression, if rhs expr is null
         if (varNode.expr == null) {
-            varNode.expr = getInitExpr(varNode.type);
+            if (varNode.type instanceof BStreamType) {
```
Comment (sameera-jayasoma) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Why do we need to special case streamType here? Put this logic inside the getInitExpr method.

Instanceof check is ugly.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 2]
```diff
@@ -189,15 +193,6 @@
 
     private Map<BLangBlockStmt, SymbolEnv> blockStmtEnvMap = new HashMap<>();
 
-    public static SemanticAnalyzer getInstance(CompilerContext context) {
```
Comment (sameera-jayasoma) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Why did you change the order of the getInstance method? Please revert this change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 3]
```diff
@@ -1221,6 +1225,187 @@ public void visit(BLangForever foreverStatement) {
...
+        }
+
+        //Validate output attribute names with stream/struct
+        for (StreamingQueryStatementNode streamingQueryStatement : foreverStatement.getStreamingQueryStatements()) {
+            checkOutputAttributes((BLangStatement) streamingQueryStatement);
+            validateOutputAttributeTypes((BLangStatement) streamingQueryStatement);
+        }
+    }
+
+    private void checkOutputAttributes(BLangStatement streamingQueryStatement) {
```
Comment (sameera-jayasoma) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Please move all your private method to the end of the class.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 2 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: Sorry, this happened when resolving a merge conflict. Revered now.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java (hunk 3 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java]: I have moved all *my* private methods.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 1 above)
Comment (mohanvive) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: We had to special case the stream since we do want to get the identifier name as well as the corresponding BStream reference to deal with Siddhi. @MaryamZi , can u please check this.

Comment (sameera-jayasoma) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Variable names only used for debugging purposes. Why do you need to store variable name? We anyway store names in the balx

Comment (sameera-jayasoma) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: I will merge for now. Lets get this fixed soon.

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: Moved stream specific logic to the getInitExpr method with https://github.com/ballerina-platform/ballerina-lang/pull/6712.
